<!DOCTYPE html>
<html>
  <head>
    <title>Alexej Gossmann</title>

        <meta charset="utf-8" />
    <meta content='text/html; charset=utf-8' http-equiv='Content-Type'>
    <meta http-equiv='X-UA-Compatible' content='IE=edge'>
    <meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1.0'>

    
    <meta name="description" content="PhD student with focus on statistics, machine learning, and programming, among other things">
    <meta property="og:description" content="PhD student with focus on statistics, machine learning, and programming, among other things" />
    
    <meta name="author" content="Alexej Gossmann" />

    

    <!--[if lt IE 9]>
      <script src="http://html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->

    <link rel="stylesheet" type="text/css" href="/style.css" />
    <link rel="alternate" type="application/rss+xml" title="Alexej Gossmann - PhD student with focus on statistics, machine learning, and programming, among other things" href="/feed.xml" />

    <!-- Created with Jekyll Now - http://github.com/barryclark/jekyll-now -->

    <!-- MathJax integration and configuration-->
    <!-- Use standard LaTeX delimiters and turn on equation numbering -->
    <script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      tex2jax: {
        inlineMath: [ ["$","$"] ],
        displayMath: [ ["$$","$$"] ],
        processEscapes: true
      },
      TeX: {
        equationNumbers: { autoNumber: "AMS" },
        Macros: {
          subscript: ['_{#1}', 1],
          superscript: ['^{#1}', 1]
        }
      }
    });
    </script>
    <script type="text/javascript" async
      src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS_HTML-full">
    </script>
 </head>

  <body>
    <div class="wrapper-masthead">
      <header class="masthead clearfix">
        <a href="/" class="site-avatar"><img src="/images/avatar.png" /></a>

        <div class="site-info">
          <h1 class="site-name"><a href="/">Alexej Gossmann</a></h1>
          <p class="site-description">PhD student with focus on statistics, machine learning, and programming, among other things</p>
        </div>

        <nav>
          <a href="/">/blog/</a>
          <a href="/publications">/publications/</a>
          <a href="/code">/code/</a>
          <a href="/about">/about/</a>
        </nav>
      </header>
    </div>

    <div id="main" role="main" class="container">
      <div class="posts">
  <!-- This loops through the paginated posts -->
  
    <article class="post">

      <h1><a href="/nmatrix-on-my-universities-hpc/">NMatrix with Intel MKL on my university's HPC</a></h1>

      <div class="entry">
        <p>In order to use <a href="https://github.com/SciRuby/nmatrix">NMatrix</a> for the statistical analysis of big genomic data, I decided to install it on my university’s high performance computing system (HPC). It is called <a href="http://crsc.tulane.edu/">Cypress</a> (like the <a href="http://imgc.allpostersimages.com/images/P-473-488-90/64/6420/5OV9100Z/posters/paul-souders-cypress-reflected-in-bayou-along-highway-61-on-stormy-summer-afternoon-new-orleans-louisiana-usa.jpg">typical New Orleans tree</a>), and it’s currently the 10th best among all American universities.</p>


      </div>

      <a href="/nmatrix-on-my-universities-hpc/" class="read-more">Read More</a>
    </article>
  
    <article class="post">

      <h1><a href="/gsoc-2015-mixed_models/">Statistical linear mixed models in Ruby with mixed_models (GSoC2015)</a></h1>

      <div class="entry">
        <p>Google Summer of Code 2015 is coming to an end. During this summer, I have learned too many things to list here about statistical modeling, Ruby and software development in general, and I had a lot of fun in the process!</p>


      </div>

      <a href="/gsoc-2015-mixed_models/" class="read-more">Read More</a>
    </article>
  
    <article class="post">

      <h1><a href="/bootstap_confidence_intervals/">Bootstrapping and bootstrap confidence intervals for linear mixed models</a></h1>

      <div class="entry">
        <p>(<strong>EDIT:</strong> I have also written <a href="http://www.alexejgossmann.com/Lehmanns_TSH_and_TPE/bootstrap_intervals_for_LMM/">a more theoretical blog post</a> on the topic.)</p>


      </div>

      <a href="/bootstap_confidence_intervals/" class="read-more">Read More</a>
    </article>
  
    <article class="post">

      <h1><a href="/mixed_models_applied_to_family_SNP_data/">A (naive) application of linear mixed models to genetics</a></h1>

      <div class="entry">
        <p>The following shows an application of class <code class="highlighter-rouge">LMM</code> from the Ruby gem <a href="https://github.com/agisga/mixed_models.git"><code class="highlighter-rouge">mixed_models</code></a> to SNP data (<a href="https://en.wikipedia.org/wiki/Single-nucleotide_polymorphism">single-nucleotide polymorphism</a>) with known pedigree structures. The family information is prior knowledge that we can model in the random effects of a linear mixed effects model.</p>


      </div>

      <a href="/mixed_models_applied_to_family_SNP_data/" class="read-more">Read More</a>
    </article>
  
    <article class="post">

      <h1><a href="/MixedModels_p_values_and_CI/">P-values and confidence intervals</a></h1>

      <div class="entry">
        <p>A few days ago I started working on hypotheses tests and confidence intervals for my project <a href="https://github.com/agisga/mixed_models"><code class="highlighter-rouge">mixed_models</code></a>, and I got pretty surprised by certain things.</p>


      </div>

      <a href="/MixedModels_p_values_and_CI/" class="read-more">Read More</a>
    </article>
  
    <article class="post">

      <h1><a href="/MixedModels_from_formula/">MixedModels Formula Interface and Categorical Variables</a></h1>

      <div class="entry">
        <p>I made some more progress on my <a href="https://github.com/agisga/MixedModels">Google Summer of Code project MixedModels</a>. The linear mixed models fitting method is now capable of handling non-numeric (i.e., categorical) predictor variables, as well as interaction effects. Moreover, I gave the method a user friendly R-formula-like interface. I will present these new capabilities of the Ruby gem with an example. Then I will briefly describe their implementation.</p>


      </div>

      <a href="/MixedModels_from_formula/" class="read-more">Read More</a>
    </article>
  
    <article class="post">

      <h1><a href="/LMM-model-specification/">Model specification for linear mixed model</a></h1>

      <div class="entry">
        <p>Last week I wrote about my implementation of an algorithm that fits a linear mixed model in Ruby using the gem <a href="https://github.com/agisga/MixedModels">MixedModels</a>, that I am working on right now. See, <a href="http://agisga.github.io/First-linear-mixed-model-fit/">first rudimentary LMM fit</a>.</p>


      </div>

      <a href="/LMM-model-specification/" class="read-more">Read More</a>
    </article>
  
    <article class="post">

      <h1><a href="/First-linear-mixed-model-fit/">A rudimentary first linear mixed model fit</a></h1>

      <div class="entry">
        <p>During the last two weeks I made some progress on my <a href="https://github.com/agisga/MixedModels">Google Summer of Code project</a>.
The Ruby gem is now capable of fitting linear mixed models. In this short blog post I want to give an example, and compare the results I get in Ruby to those obtained by <code class="highlighter-rouge">lme4</code> in R.</p>


      </div>

      <a href="/First-linear-mixed-model-fit/" class="read-more">Read More</a>
    </article>
  
    <article class="post">

      <h1><a href="/Dissect_lmer_part3/">Dissecting lme4's lmer function. Part 3.</a></h1>

      <div class="entry">
        <p>This is the final part of my analysis of the function <code class="highlighter-rouge">lmer</code>, which is used to fit linear mixed models in the R package <code class="highlighter-rouge">lme4</code>. In two previous blog posts, we have seen the general layout of the function <code class="highlighter-rouge">lmer</code>, the dealings with the R model formula, and the setting up of the objective function for the optimization (see <a href="http://agisga.github.io/Dissect_lmer_part1/">part 1</a> and <a href="http://agisga.github.io/Dissect_lmer_part2/">part 2</a>).</p>


      </div>

      <a href="/Dissect_lmer_part3/" class="read-more">Read More</a>
    </article>
  
    <article class="post">

      <h1><a href="/Dissect_lmer_part2/">Dissecting lme4's lmer function. Part 2.</a></h1>

      <div class="entry">
        <p><a href="http://agisga.github.io/Dissect_lmer_part1/">Last time</a> I started to analyze the function <code class="highlighter-rouge">lmer</code> that is used to fit linear mixed models in the R package <code class="highlighter-rouge">lme4</code>. I have delineated the general steps taken by <code class="highlighter-rouge">lmer</code>, and looked at the employed formula module in more detail. The formula module evaluates the provided R model formula to model matrices, vectors and parameters. The next step is to use these to define the objective function that needs to be minimized, which is the profiled deviance or the profiled REML criterion in this case. The objective function is returned by the function <code class="highlighter-rouge">mkLmerDevfun</code> which is dissected in what follows.</p>


      </div>

      <a href="/Dissect_lmer_part2/" class="read-more">Read More</a>
    </article>
  
</div>

<!-- Pagination links -->
<div class="pagination">
  
    <a href="/page2" class="previous">Previous</a>
  
  <span class="page_number ">Page: 3 of 4</span>
  
    <a href="/page4" class="next">Next</a>
  
</div>

    </div>

    <div class="wrapper-footer">
      <div class="container">
        <footer class="footer">
          
<a href="mailto:alexej.go@gmail.com"><i class="svg-icon email"></i></a>
<a href="https://www.facebook.com/alexej.yexela"><i class="svg-icon facebook"></i></a>

<a href="https://github.com/agisga"><i class="svg-icon github"></i></a>

<a href="https://www.linkedin.com/in/alexejgossmann"><i class="svg-icon linkedin"></i></a>

<a href="/feed.xml"><i class="svg-icon rss"></i></a>




        </footer>
      </div>
    </div>

    
	<!-- Google Analytics -->
	<script>
		(function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
		(i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
		m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
		})(window,document,'script','//www.google-analytics.com/analytics.js','ga');

		ga('create', 'UA-94080131-1', 'auto');
		ga('send', 'pageview', {
		  'page': '/page3/',
		  'title': ''
		});
	</script>
	<!-- End Google Analytics -->


  </body>
</html>
