<!DOCTYPE html>
<html>
  <head>
    <title>Alexej Gossmann â€“ I am interested in math, stats, coding, genetics, ...</title>

        <meta charset="utf-8" />
    <meta content='text/html; charset=utf-8' http-equiv='Content-Type'>
    <meta http-equiv='X-UA-Compatible' content='IE=edge'>
    <meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1.0'>

    
    <meta name="description" content="I am interested in math, stats, coding, genetics, ...">
    <meta property="og:description" content="I am interested in math, stats, coding, genetics, ..." />
    
    <meta name="author" content="Alexej Gossmann" />

    

    <!--[if lt IE 9]>
      <script src="http://html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->

    <link rel="stylesheet" type="text/css" href="/style.css" />
    <link rel="alternate" type="application/rss+xml" title="Alexej Gossmann - I am interested in math, stats, coding, genetics, ..." href="/feed.xml" />

    <!-- Created with Jekyll Now - http://github.com/barryclark/jekyll-now -->
 
    <!-- MathJax interation-->
    <script type="text/x-mathjax-config">
      MathJax.Hub.Config({tex2jax: {inlineMath: [["$","$"]]}});
      MathJax.Hub.Config({TeX: {Macros:{subscript:['_{#1}',1],superscript:['^{#1}',1]}}});
    </script> 
    <!-- Turn on equation numbering -->
    <script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      TeX: { equationNumbers: { autoNumber: "AMS" } }
    });
    </script>
    <script type="text/javascript"
      src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS_HTML-full">
    </script>
 </head>

  <body>
    <div class="wrapper-masthead">
      <header class="masthead clearfix">
        <a href="/" class="site-avatar"><img src="https://avatars0.githubusercontent.com/u/11449372?v=3&s=460" /></a>

        <div class="site-info">
          <h1 class="site-name"><a href="/">Alexej Gossmann</a></h1>
          <p class="site-description">I am interested in math, stats, coding, genetics, ...</p>
        </div>

        <nav>
          <a href="/about" class="blue">About</a>
          <a href="/" class="yellow">Blog</a>
          <a href="/talks" class="green">Presentations</a>
          <a href="/publications" class="magenta">Publications</a>
          <a href="/software" class="cyan">Software</a>
        </nav>
      </header>
    </div>

    <div id="main" role="main" class="container">
      <div class="posts">
  <!-- This loops through the paginated posts -->
  
    <article class="post">

      <h1><a href="/bootstap_confidence_intervals/">Bootstrapping and bootstrap confidence intervals for linear mixed models</a></h1>

      <div class="entry">
        <p>During the last couple of days, I have added some parametric bootstrap capabilities to <a href="https://github.com/agisga/mixed_models.git"><code>mixed_models</code></a>. The following demonstrates first how to resample fixed effects coefficient estimates via bootstrapping. Then various types of bootstrap confidence intervals are presented. Examples of computation and a brief comparison of different types of bootstrap confidence intervals are given.</p>

      </div>

      <a href="/bootstap_confidence_intervals/" class="read-more">Read More</a>
    </article>
  
    <article class="post">

      <h1><a href="/mixed_models_applied_to_family_SNP_data/">A (naive) application of linear mixed models to genetics</a></h1>

      <div class="entry">
        <p>The following shows an application of class <code>LMM</code> from the Ruby gem <a href="https://github.com/agisga/mixed_models.git"><code>mixed_models</code></a> to SNP data (<a href="https://en.wikipedia.org/wiki/Single-nucleotide_polymorphism">single-nucleotide polymorphism</a>) with known pedigree structures. The family information is prior knowledge that we can model in the random effects of a linear mixed effects model.</p>

      </div>

      <a href="/mixed_models_applied_to_family_SNP_data/" class="read-more">Read More</a>
    </article>
  
    <article class="post">

      <h1><a href="/MixedModels_p_values_and_CI/">P-values and confidence intervals</a></h1>

      <div class="entry">
        <p>A few days ago I started working on hypotheses tests and confidence intervals for my project <a href="https://github.com/agisga/mixed_models"><code>mixed_models</code></a>, and I got pretty surprised by certain things.</p>

      </div>

      <a href="/MixedModels_p_values_and_CI/" class="read-more">Read More</a>
    </article>
  
</div>

<!-- Pagination links -->
<div class="pagination">
  
    <a href="/index.html" class="previous">Previous</a>
  
  <span class="page_number ">Page: 2 of 6</span>
  
    <a href="/page3" class="next">Next</a>
  
</div>

<!--<div class="posts">
  
    <article class="post">

      <h1><a href="/TSH_and_TPE_observations/">"Testing Statistical Hypotheses" and "Theory of Point Estimation" impressions</a></h1>

      <div class="entry">
        <p>I spent much of the last two months reading Lehmann &amp; Romano &quot;Testing Statistical Hypotheses&quot; (3rd ed.) and Lehmann &amp; Casella &quot;Theory of Point Estimation&quot; (2nd ed.), abbr. TSH and TPE. The following is a collection of <del>random facts</del> observations I made while reading TSH and TPE. The choice of topics is biased towards application in regression models.</p>

      </div>

      <a href="/TSH_and_TPE_observations/" class="read-more">Read More</a>
    </article>
  
    <article class="post">

      <h1><a href="/nmatrix-on-my-universities-hpc/">NMatrix with Intel MKL on my university's HPC</a></h1>

      <div class="entry">
        <p>In order to use <a href="https://github.com/SciRuby/nmatrix">NMatrix</a> for the statistical analysis of big genomic data, I decided to install it on my university&#39;s high performance computing system (HPC). It is called <a href="http://crsc.tulane.edu/">Cypress</a> (like the <a href="http://imgc.allpostersimages.com/images/P-473-488-90/64/6420/5OV9100Z/posters/paul-souders-cypress-reflected-in-bayou-along-highway-61-on-stormy-summer-afternoon-new-orleans-louisiana-usa.jpg">typical New Orleans tree</a>), and it&#39;s currently the 10th best among all American universities. </p>

      </div>

      <a href="/nmatrix-on-my-universities-hpc/" class="read-more">Read More</a>
    </article>
  
    <article class="post">

      <h1><a href="/gsoc-2015-mixed_models/">Statistical linear mixed models in Ruby with mixed_models (GSoC2015)</a></h1>

      <div class="entry">
        <p>Google Summer of Code 2015 is coming to an end. During this summer, I have learned too many things to list here about statistical modeling, Ruby and software development in general, and I had a lot of fun in the process!</p>

      </div>

      <a href="/gsoc-2015-mixed_models/" class="read-more">Read More</a>
    </article>
  
    <article class="post">

      <h1><a href="/bootstap_confidence_intervals/">Bootstrapping and bootstrap confidence intervals for linear mixed models</a></h1>

      <div class="entry">
        <p>During the last couple of days, I have added some parametric bootstrap capabilities to <a href="https://github.com/agisga/mixed_models.git"><code>mixed_models</code></a>. The following demonstrates first how to resample fixed effects coefficient estimates via bootstrapping. Then various types of bootstrap confidence intervals are presented. Examples of computation and a brief comparison of different types of bootstrap confidence intervals are given.</p>

      </div>

      <a href="/bootstap_confidence_intervals/" class="read-more">Read More</a>
    </article>
  
    <article class="post">

      <h1><a href="/mixed_models_applied_to_family_SNP_data/">A (naive) application of linear mixed models to genetics</a></h1>

      <div class="entry">
        <p>The following shows an application of class <code>LMM</code> from the Ruby gem <a href="https://github.com/agisga/mixed_models.git"><code>mixed_models</code></a> to SNP data (<a href="https://en.wikipedia.org/wiki/Single-nucleotide_polymorphism">single-nucleotide polymorphism</a>) with known pedigree structures. The family information is prior knowledge that we can model in the random effects of a linear mixed effects model.</p>

      </div>

      <a href="/mixed_models_applied_to_family_SNP_data/" class="read-more">Read More</a>
    </article>
  
    <article class="post">

      <h1><a href="/MixedModels_p_values_and_CI/">P-values and confidence intervals</a></h1>

      <div class="entry">
        <p>A few days ago I started working on hypotheses tests and confidence intervals for my project <a href="https://github.com/agisga/mixed_models"><code>mixed_models</code></a>, and I got pretty surprised by certain things.</p>

      </div>

      <a href="/MixedModels_p_values_and_CI/" class="read-more">Read More</a>
    </article>
  
    <article class="post">

      <h1><a href="/MixedModels_from_formula/">MixedModels Formula Interface and Categorical Variables</a></h1>

      <div class="entry">
        <p>I made some more progress on my <a href="https://github.com/agisga/MixedModels">Google Summer of Code project MixedModels</a>.
The linear mixed models fitting method is now capable of handling non-numeric (i.e. categorical) predictor variables, as well as interaction effects. Moreover, I gave the method a user friendly R-formula-like interface.
I will present these new capabilities of the Ruby gem with an example. Then I will briefly describe their implementation.</p>

      </div>

      <a href="/MixedModels_from_formula/" class="read-more">Read More</a>
    </article>
  
    <article class="post">

      <h1><a href="/LMM-model-specification/">Model specification for linear mixed model</a></h1>

      <div class="entry">
        <p>Last week I wrote about my implementation of an algorithm that fits a linear mixed model in Ruby using the gem <a href="https://github.com/agisga/MixedModels">MixedModels</a>, that I am working on right now. See, <a href="http://agisga.github.io/First-linear-mixed-model-fit/">first rudimentary LMM fit</a>.</p>

      </div>

      <a href="/LMM-model-specification/" class="read-more">Read More</a>
    </article>
  
    <article class="post">

      <h1><a href="/First-linear-mixed-model-fit/">A rudimentary first linear mixed model fit</a></h1>

      <div class="entry">
        <p>During the last two weeks I made some progress on my <a href="https://github.com/agisga/MixedModels">Google Summer of Code project</a>.
The Ruby gem is now capable of fitting linear mixed models.
In this short blog post I want to give an example, and compare the results I get in Ruby to those obtained by <code>lme4</code> in R.</p>

      </div>

      <a href="/First-linear-mixed-model-fit/" class="read-more">Read More</a>
    </article>
  
    <article class="post">

      <h1><a href="/Dissect_lmer_part3/">Dissecting lme4's lmer function. Part 3.</a></h1>

      <div class="entry">
        <p>This is the final part of my analysis of the function <code>lmer</code>, which is used to fit linear mixed models in the R package <code>lme4</code>.
In two previous blog posts, we have seen the general layout of the function <code>lmer</code>, the dealings with the R model formula, and the setting up of the objective function for the optimization (see <a href="http://agisga.github.io/Dissect_lmer_part1/">part 1</a> and <a href="http://agisga.github.io/Dissect_lmer_part2/">part 2</a>).</p>

      </div>

      <a href="/Dissect_lmer_part3/" class="read-more">Read More</a>
    </article>
  
    <article class="post">

      <h1><a href="/Dissect_lmer_part2/">Dissecting lme4's lmer function. Part 2.</a></h1>

      <div class="entry">
        <p><a href="http://agisga.github.io/Dissect_lmer_part1/">Last time</a> I started to analyze the function <code>lmer</code> that is used to fit linear mixed models in the R package <code>lme4</code>.
I have delineated the general steps taken by <code>lmer</code>, and looked at the employed formula module in more detail. The formula module evaluates the provided R model formula to model matrices, vectors and parameters. The next step is to use these to define the objective function that needs to be minimized, which is the profiled deviance or the profiled REML criterion in this case. The objective function is returned by the function <code>mkLmerDevfun</code> which is dissected in what follows.</p>

      </div>

      <a href="/Dissect_lmer_part2/" class="read-more">Read More</a>
    </article>
  
    <article class="post">

      <h1><a href="/Dissect_lmer_part1/">Dissecting lme4's lmer function. Part 1.</a></h1>

      <div class="entry">
        <p>This blog posts marks the start of my <a href="https://www.google-melange.com/gsoc/homepage/google/gsoc2015">Google Summer of Code</a> project with the <a href="http://sciruby.com">Ruby Science Foundation</a>, where I will develop mixed linear models software for Ruby. As a preparation for my GSoC project, I will dedicate a couple of blog posts to a meticulous analysis of <a href="https://github.com/lme4/lme4.git"><code>lme4</code></a> code (so that I can steal all the ideas from it!).</p>

      </div>

      <a href="/Dissect_lmer_part1/" class="read-more">Read More</a>
    </article>
  
    <article class="post">

      <h1><a href="/BVP/">Solve two-point boundary value problems in Ruby with spitzy</a></h1>

      <div class="entry">
        <p>A few days ago I programmed a numerical method for the solution for two-point boundary value problems, and today I discovered that I can use <a href="https://www.mathjax.org/">MathJax</a> to display mathematical formulas in here (although there are some inconveniences related to the use of underscores). So, here goes another blog post!</p>

      </div>

      <a href="/BVP/" class="read-more">Read More</a>
    </article>
  
    <article class="post">

      <h1><a href="/ODE/">Solve ODEs in Ruby with spitzy</a></h1>

      <div class="entry">
        <p>Over the weekend I have written a couple of numerical solvers for one-dimensional <a href="http://en.wikipedia.org/wiki/Initial_value_problem">initial value problems</a> in Ruby, and added them to my project <a href="https://github.com/agisga/spitzy">spitzy</a>.</p>

      </div>

      <a href="/ODE/" class="read-more">Read More</a>
    </article>
  
    <article class="post">

      <h1><a href="/Advection-Equation/">Solve an advection equation in Ruby with spitzy</a></h1>

      <div class="entry">
        <p>A couple of days ago I started working on a collection of numerical methods for differential equations, wirtten in pure Ruby (I have conviced the professor of my numerical DE class that that`s a good idea for my final project in said class).</p>

      </div>

      <a href="/Advection-Equation/" class="read-more">Read More</a>
    </article>
  
    <article class="post">

      <h1><a href="/Permute-Columns-Of-An-NMatrix/">NMatrix column permutations</a></h1>

      <div class="entry">
        <p>Recently I got surprised by the behaviour of <code>#permute_columns</code> in the <em>Ruby</em> gem <em>NMatrix</em>.</p>

      </div>

      <a href="/Permute-Columns-Of-An-NMatrix/" class="read-more">Read More</a>
    </article>
  
    <article class="post">

      <h1><a href="/Hello-World/">Hello World!</a></h1>

      <div class="entry">
        <p>Hello World!</p>

      </div>

      <a href="/Hello-World/" class="read-more">Read More</a>
    </article>
  
  </div>-->

    </div>

    <div class="wrapper-footer">
      <div class="container">
        <footer class="footer">
          



<a href="http://github.com/agisga"><i class="svg-icon github"></i></a>

<a href="http://linkedin.com/in/alexejgossmann"><i class="svg-icon linkedin"></i></a>


<a href="http://twitter.com/agisga"><i class="svg-icon twitter"></i></a>


        </footer>
      </div>
    </div>

    

  </body>
</html>
